src/AdvancedVI.jl

```diff
@@ -210,61 +161,53 @@ Iteratively updates parameters by calling `grad!` and using the given `optimizer
 the steps.
 """
 function optimize!(
-    vo,
-    alg::VariationalInference,
-    q,
-    model,
-    θ::AbstractVector{<:Real};
-    optimizer = TruncatedADAGrad()
+    grad_estimator::AbstractGradientEstimator,
+    rebuild::Function,
+    ℓπ::Function,
+    n_max_iter::Int,
+    λ::AbstractVector{<:Real};
+    optimizer = TruncatedADAGrad(),
+    rng       = Random.GLOBAL_RNG
 )
-    # TODO: should we always assume `samples_per_step` and `max_iters` for all algos?
-    alg_name = alg_str(alg)
-    samples_per_step = alg.samples_per_step
-    max_iters = alg.max_iters
-    
-    num_params = length(θ)
+    obj_name = objective(grad_estimator)
 
     # TODO: really need a better way to warn the user about potentially
     # not using the correct accumulator
-    if (optimizer isa TruncatedADAGrad) && (θ ∉ keys(optimizer.acc))
+    if (optimizer isa TruncatedADAGrad) && (λ ∉ keys(optimizer.acc))
         # this message should only occurr once in the optimization process
-        @info "[$alg_name] Should only be seen once: optimizer created for θ" objectid(θ)
+        @info "[$obj_name] Should only be seen once: optimizer created for θ" objectid(λ)
     end
 
-    diff_result = DiffResults.GradientResult(θ)
+    grad_buf = DiffResults.GradientResult(λ)
 
     i = 0
-    prog = if PROGRESS[]
-        ProgressMeter.Progress(max_iters, 1, "[$alg_name] Optimizing...", 0)
-    else
-        0
-    end
+    prog = ProgressMeter.Progress(
+        n_max_iter; desc="[$obj_name] Optimizing...", barlen=0, enabled=PROGRESS[])
 
     # add criterion? A running mean maybe?
-    time_elapsed = @elapsed while (i < max_iters) # & converged
-        grad!(vo, alg, q, model, θ, diff_result, samples_per_step)
-
-        # apply update rule
-        Δ = DiffResults.gradient(diff_result)
-        Δ = apply!(optimizer, θ, Δ)
-        @. θ = θ - Δ
+    time_elapsed = @elapsed begin
+        for i = 1:n_max_iter
+            stats = estimate_gradient!(rng, grad_estimator, λ, rebuild, ℓπ, grad_buf)
+            
+            # apply update rule
+            Δλ = DiffResults.gradient(grad_buf)
+            Δλ = apply!(optimizer, λ, Δλ)
+            @. λ = λ - Δλ
+
+            stat′ = (Δλ=norm(Δλ),)
+            stats = merge(stats, stat′)
```

Comment:
What is the reasoning for keeping track of \Delta\lambda ?

Reply:
I left that in because the current codebase [does it](https://github.com/TuringLang/AdvancedVI.jl/blob/53540e51dbe4de09ebba0808e1b353cd41f1209c/src/AdvancedVI.jl#L252). I assume it was intended as a simple conversion diagnostic?

Reply:
We're talking about the `norm`, right? If so, it's because it's useful to get a sense of convergence, no?